### PR TITLE
[build][option] Introduce `OPENCV_DISABLE_THREAD_SUPPORT` option.

### PR DESCRIPTION
**Pull request by fpetrogalli**
The option forces the library to build without the use of `#include
<thread>`. The modules for which this is not possible has been
conditionally disabled when the option is active.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [X] I agree to contribute to the project under Apache 2 License.
- [X] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [ ] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

```
force_builders=linux,docs,ARMv8,ARMv7,Custom
build_image:Custom=javascript
buildworker:Custom=linux-4,linux-6
```

### CODE REVIEW

**Comment by fpetrogalli** [comment]
@alalek, @vpisarev - I have factored out the option `CV_DISABLE_MULTI_THREAD` as suggested from https://github.com/opencv/opencv/pull/19921/

**Comment by fpetrogalli** [comment]
HI @alalek - I am taking the comment at https://github.com/opencv/opencv/pull/19921#issuecomment-827096736 here as it seems specific to this PR.

> I believe stubs should be created in the #else branch to handle these unsupported cases (which throws CV_Error(Error::StsNotImplemented, ...))

Are you saying that code that is depending on `#include <thread>` should be modified as follows?

```
#ifndef CV_DISABLE_MULTI_THREAD
#include <thread>
#else 
// Stub implementation of whatever is needed from pthread, that raises errors for unimplemented methods:
namespace std {
  struct thread {
    // ...
    void join() {
      CV_Error(Error::StsNotImplemented, ...);
    }
  };
}
#endif
```

If that's the case, I think it would make sense to merge the stubs in the `#else` branch in a common header file. What would be the best place to store such file for automatic inclusion in all the project? Probably the main include of the `code` module?

**Comment by alalek** [comment]
That comment is about stubs for OpenCV types, like `cv::AsyncArray`, `cv::Mutex` (exception for lock()/unlock() only), `cv::AutoLock` (no-op instead of exception).
Creating stubs for `std::thread` looks dangerous.

**Comment by fpetrogalli** [comment]
> That comment is about stubs for OpenCV types, like `cv::AsyncArray`, `cv::Mutex` (exception for lock()/unlock() only), `cv::AutoLock` (no-op instead of exception).

I see - thank you for explaining @alalek . Are there any other class I should be aware of that I should conditionally redefined based on the macro `CV_DISABLE_MULTI_THREAD`?

> Creating stubs for std::thread looks dangerous.

Yep, but I thought that was wanted me to do with the "port" file that you wanted me to create in https://github.com/opencv/opencv/pull/19921/ I didn't get as far as implementing the rest of the `struct thread` there because many things are disabled automatically when using a baremetal toolchain:

```
-- To be built: core features2d highgui imgcodecs imgproc ml photo video
-- Disabled: flann ts videoio world
-- Disabled by dependency: calib3d objdetect stitching
-- Unavailable: dnn gapi java python2 python3```

If we want the option `CV_DISABLE_MULTI_THREAD` to be an option for the whole library and (understandably) we don't want to create stubs for `std::thread`, I think I'll have to disable all the modules that do `#include <thread>` (either directly or indirectly via the use of `<future>` or other things). As far as I can tell, these should be at least the ones listed in the `Unavailable` field.

**Comment by fpetrogalli** [comment]
@vpisarev  - is the arm64 builder at https://build.opencv.org.cn/job/precommit/job/ubuntu-20.04-arm64/ broken? Or is the failure of the latest commit a genuine failure caused by my commit?

**Comment by berak** [comment]
@fpetrogalli  -- you're safe to ignore the failing `OpenCV CN Ubuntu 18.04 x86-64` jenkins cn bot

(it fails on a java test, unrelated to your commit, and it does that *every time*, it's an infrastructure problem, not your fault)

**Comment by fpetrogalli** [comment]
@vpisarev @alalek  - gentle ping on this patch too. The OpenCV CN CI failures seem to be unrelated too as for #20010

**Comment by fpetrogalli** [comment]
@alalek @vpisarev - may I kindly ask some feedback on this patch? I believe I addressed all concerns raised up to this point.

Kind regards,

Francesco

File: CMakeLists.txt [hunk 1]
```diff
@@ -523,6 +525,32 @@ if(OPENCV_DISABLE_FILESYSTEM_SUPPORT)
   add_definitions(-DOPENCV_HAVE_FILESYSTEM_SUPPORT=0)
 endif()
 
+if(CV_DISABLE_MULTI_THREAD)
+  add_definitions(-DCV_DISABLE_MULTI_THREAD)
+  # This is required to disable threads in the ts module, as
+  # described in `ts_gtest.h`.
+  add_definitions(-DGTEST_HAS_PTHREAD=0)
```
Comment (alalek) [inline CMakeLists.txt:532]: I believe root CMake file should not know details of `ts` module.
Please move this into `modules/ts/CMakeLists.txt`, use

```
ocv_target_compile_definitions(${the_module} PUBLIC GTEST_HAS_PTHREAD=0)
```

instead.

File: 3rdparty/libwebp/CMakeLists.txt [hunk 2]
```diff
@@ -32,7 +32,9 @@ endif()
 #         Define the library target:
 # ----------------------------------------------------------------------------------
 
-add_definitions(-DWEBP_USE_THREAD)
+if (NOT CV_DISABLE_MULTI_THREAD)
```
Comment (alalek) [inline 3rdparty/libwebp/CMakeLists.txt:35]: `if(` without space.

File: modules/core/CMakeLists.txt [hunk 3]
```diff
@@ -83,6 +83,10 @@ ocv_glob_module_sources(SOURCES "${OPENCV_MODULE_opencv_core_BINARY_DIR}/version
 ocv_module_include_directories(${the_module} ${ZLIB_INCLUDE_DIRS} ${OPENCL_INCLUDE_DIRS})
 if(ANDROID AND HAVE_CPUFEATURES)
   ocv_append_source_file_compile_definitions(${CMAKE_CURRENT_SOURCE_DIR}/src/system.cpp "HAVE_CPUFEATURES=1")
+  if(CV_DISABLE_MULTI_THREAD)
+    ocv_append_source_file_compile_definitions(${CMAKE_CURRENT_SOURCE_DIR}/src/system.cpp
+      "CV_DISABLE_MULTI_THREAD")
```
Comment (alalek) [inline modules/core/CMakeLists.txt:88]: global `add_definition` should be enough.

File: modules/core/src/utils/logtagmanager.hpp [hunk 4]
```diff
@@ -37,8 +37,13 @@ class LogTagManager
     // also, extensible functions (accepting user-provided callback) are not allowed
     // to call LogTagManger (to prevent iterator invalidation), which needs enforced
     // with a non-recursive mutex.
+#ifndef CV_DISABLE_MULTI_THREAD
     using MutexType = std::mutex;
     using LockType = std::lock_guard<MutexType>;
+#else // CV_DISABLE_MULTI_THREAD
+    using MutexType = cv::Mutex;
+    using LockType = cv::AutoLock;
```
Comment (alalek) [inline modules/core/src/utils/logtagmanager.hpp:45]: Perhaps we can use this branch only with `cv::` classes.

File: CMakeLists.txt [hunk 5]
```diff
@@ -514,10 +514,38 @@ OCV_OPTION(OPENCV_ENABLE_MEMALIGN     "Enable posix_memalign or memalign usage"
 OCV_OPTION(ENABLE_PYLINT              "Add target with Pylint checks"                            (BUILD_DOCS OR BUILD_EXAMPLES) IF (NOT CMAKE_CROSSCOMPILING AND NOT APPLE_FRAMEWORK) )
 OCV_OPTION(ENABLE_FLAKE8              "Add target with Python flake8 checker"                    (BUILD_DOCS OR BUILD_EXAMPLES) IF (NOT CMAKE_CROSSCOMPILING AND NOT APPLE_FRAMEWORK) )
 
+OCV_OPTION(CV_DISABLE_MULTI_THREAD    "Build the library without multi-threaded code."            OFF)
```
[the remark this answers is not in the document]
Comment (fpetrogalli) [reply CMakeLists.txt:518]: Done

File: CMakeLists.txt (hunk 1 above)
Comment (fpetrogalli) [reply CMakeLists.txt:532]: Done

File: 3rdparty/libwebp/CMakeLists.txt (hunk 2 above)
Comment (fpetrogalli) [reply 3rdparty/libwebp/CMakeLists.txt:35]: Done.

File: modules/core/CMakeLists.txt (hunk 3 above)
Comment (fpetrogalli) [reply modules/core/CMakeLists.txt:88]: Done.

File: modules/core/src/utils/logtagmanager.hpp (hunk 4 above)
Comment (fpetrogalli) [reply modules/core/src/utils/logtagmanager.hpp:45]: Done

**Comment by fpetrogalli** [comment]
Thank you for the review, @alalek !

File: modules/core/include/opencv2/core/utility.hpp [hunk 6]
```diff
@@ -714,9 +714,27 @@ void Mat::forEach_impl(const Functor& operation) {
 /////////////////////////// Synchronization Primitives ///////////////////////////////
 
 #if !defined(_M_CEE)
+#ifndef OPENCV_DISABLE_THREAD_SUPPORT
```
Comment (fpetrogalli) [inline modules/core/include/opencv2/core/utility.hpp:717]: @alalek  - I was wondering, should I turn these in `#if !OPENCV_DISABLE_THREAD_SUPPORT` and set the macro to 0 by default, like for the filesystem on?

Francesco

Comment (alalek) [reply modules/core/include/opencv2/core/utility.hpp:717]: In general in makes sense (at least to catch typos through `-Wundef`)
However, we don't use 0/1 "mandatory" values on macros with "_DISABLE_" prefix.
Filesystem's macro name is `OPENCV_HAVE_FILESYSTEM_SUPPORT`.
If you want to switch then it makes sense to rename the macro too (C++ only part without CMake names)

File: CMakeLists.txt [hunk 7]
```diff
@@ -515,6 +515,8 @@ OCV_OPTION(ENABLE_PYLINT              "Add target with Pylint checks"
 OCV_OPTION(ENABLE_FLAKE8              "Add target with Python flake8 checker"                    (BUILD_DOCS OR BUILD_EXAMPLES) IF (NOT CMAKE_CROSSCOMPILING AND NOT APPLE_FRAMEWORK) )
 OCV_OPTION(OPENCV_DISABLE_FILESYSTEM_SUPPORT "Disable filesystem support" OFF)
 
+OCV_OPTION(OPENCV_DISABLE_THREAD_SUPPORT    "Build the library without multi-threaded code."            OFF)
+
```
Comment (alalek) [inline CMakeLists.txt:519]: Why are there options are placed near python linter options group?
Why is there extra spacing between them?

File: CMakeLists.txt [hunk 8]
```diff
@@ -523,6 +525,29 @@ if(OPENCV_DISABLE_FILESYSTEM_SUPPORT)
...
+  set(WITH_OPENCL OFF)
+
+  # These modules require `#include
+  # <[thread|mutex|condition_variable|future]>` and linkage into
+  # `libpthread` to work. Notice that disabling `videoio` disables
+  # the `ts` module.
+  set(BUILD_opencv_videoio OFF)
+  set(BUILD_opencv_objdetect OFF)
+  set(BUILD_opencv_gapi OFF)
+  set(BUILD_opencv_dnn OFF)
```
Comment (alalek) [inline CMakeLists.txt:548]: Comment about `GTEST_HAS_PTHREAD` is applicable for all these flags too.
Root CMake script should not know implementation details of other modules or dependencies.

File: CMakeLists.txt [hunk 9]
```diff
@@ -1458,6 +1488,15 @@ ocv_build_features_string(parallel_status EXCLUSIVE
   ELSE "none")
 status("")
 status("  Parallel framework:" "${parallel_status}")
+if (OPENCV_DISABLE_THREAD_SUPPORT)
+  status("" "Multi thread code explicitly disabled with OPENCV_DISABLE_THREAD_SUPPORT.")
+  if(HAVE_PTHREADS_PF OR HAVE_HPX OR HAVE_OPENMP OR HAVE_GCD OR HAVE_CONCURRENCY)
+    status("" "Warning: not all parallel frameworks have been disabled (using ${parallel_status}).")
```
Comment (alalek) [inline CMakeLists.txt:1494]: Why are these warnings should go into `getBuildInformation()`?

File: modules/core/src/async.cpp [hunk 10]
```diff
@@ -202,7 +202,7 @@ struct AsyncArray::Impl
     {
         if (future_is_returned && refcount_future == 0)
             CV_Error(Error::StsError, "Associated AsyncArray has been destroyed");
-#ifdef CV_CXX11
+#if defined(CV_CXX11) && !defined(OPENCV_DISABLE_THREAD_SUPPORT)
         std::unique_lock<std::mutex> lock(mtx);
 #else
         cv::AutoLock lock(mtx);
```
Comment (alalek) [inline modules/core/src/async.cpp:208]: "#else" non C++11 branch is candidate for removal.
This patch forces to keep this.

IMHO, we need dedicated implementation for `AsyncArray` for case without threads (or even forbid using of this).

File: modules/core/src/system.cpp [hunk 11]
```diff
@@ -1489,28 +1491,51 @@ void TlsAbstraction::setData_(void *pData)
...
     cv::__termination = true;  // DllMain is missing in static builds
     if (pthread_key_delete(tlsKey) != 0)
     {
         // Don't use logging here
         fprintf(stderr, "OpenCV ERROR: TlsAbstraction::~TlsAbstraction(): pthread_key_delete() call failed\n");
         fflush(stderr);
     }
+#else // OPENCV_DISABLE_THREAD_SUPPORT
+    CV_Error(cv::Error::StsNotImplemented,
+             "TlsAbstraction is disabled by OPENCV_DISABLE_THREAD_SUPPORT=ON");
```
Comment (alalek) [inline modules/core/src/system.cpp:1518]: destructors must not throw exceptions.

Introduced changes about TLS should be minimized.
Go on upper level, not need to "stub" local functions - work with public API only.

File: CMakeLists.txt (hunk 7 above)
Comment (fpetrogalli) [reply CMakeLists.txt:519]: > Why are there options are placed near python linter options group?

No meaningful reason for this. Please let me know where to move them. Do you want me to move the FILESYSTEM one in a separate patch?

> Why is there extra spacing between them?

Again, no reason for this :). I will remove the extra space after moving.

File: CMakeLists.txt (hunk 8 above)
Comment (fpetrogalli) [reply CMakeLists.txt:548]: Just to make sure I don't miss anything: the only option that is left to be moved is `OPJ_USE_THREAD`, right?

Also, am I correct in assuming that I can do the following in the openjpeg's CMakeList.txt?

```
if (OPENCV_DISABLE_THREAD_SUPPORT)
  set(OPJ_USE_THREAD OFF)
endif()
```

(Put it in other words: is `OPENCV_DISABLE_THREAD_SUPPORT` visible in the cmake configuration of openjpg/?)

File: CMakeLists.txt (hunk 9 above)
Comment (fpetrogalli) [reply CMakeLists.txt:1494]: Yeah, they are here mostly because I was using them for debugging when I was developing this patch. I am happy to remove them, or to move them to a more appropriate place. Let me know what you prefer.

File: modules/core/src/async.cpp (hunk 10 above)
Comment (fpetrogalli) [reply modules/core/src/async.cpp:208]: I am happy to provide an implementation of `AsyncArray` that does not use threads. I'll give it a go, but I might need some guidance on this.

Just to make sure I am going in the right direction. You want me to have a big `#ifundef / #else / #endif` switch where the `#else` branch exposes the curret implementation (threaded) of `AsyncArray`, while the `#ifundef OPENCV_DISABLE_THREAD_SUPPORT` will implement AsyncArray without using threads (ro shoudl it be just stubs?).

Also, what the alternative...

> even forbid using of this

 ... consist of? Disabling the use of AsyncArray in all places where it is used?

File: modules/core/src/system.cpp (hunk 11 above)
Comment (fpetrogalli) [reply modules/core/src/system.cpp:1518]: > destructors must not throw exceptions.
> 
Sure! Why? (Genuine question, I am not challenging your comment!)

> Introduced changes about TLS should be minimized.
> Go on upper level, not need to "stub" local functions - work with public API only.

What does "go on upper level" mean? Leave `TlsAbstraction` unchanges, and use the `#ifundef / #else / #endif` in all places where `TlsAbstraction` is used?

**Comment by fpetrogalli** [comment]
Thank you for the review @alalek !

File: CMakeLists.txt (hunk 7 above)
Comment (alalek) [reply CMakeLists.txt:519]: It makes sense to move them (both) upper, right after `OPENCV_ENABLE_MEMALIGN`

File: CMakeLists.txt (hunk 8 above)
Comment (alalek) [reply CMakeLists.txt:548]: Yes, this option should work.

(please remove space after `if` and `bracket`)

File: CMakeLists.txt (hunk 9 above)
Comment (alalek) [reply CMakeLists.txt:1494]: Prefer to use CMake's `message(WARNING ..)`

File: modules/core/src/async.cpp (hunk 10 above)
Comment (alalek) [reply modules/core/src/async.cpp:208]: Stub with exceptions should work for now (there are no many places which uses AsyncArray)

File: modules/core/src/system.cpp (hunk 11 above)
Comment (alalek) [reply modules/core/src/system.cpp:1518]: Main goal here is to minimize amount of `#if` conditions.

Public API for TLS is `TLSDataContainer`. Other things (like `TlsAbstraction`, `TlsStorage`) are local implementation details - we should not care about them here.
Stub `TLSDataContainer` with straightforward implementation.

File: CMakeLists.txt (hunk 8 above)
Comment (fpetrogalli) [reply CMakeLists.txt:548]: Done

File: CMakeLists.txt (hunk 9 above)
Comment (fpetrogalli) [reply CMakeLists.txt:1494]: In the same place? Or should I move it to a place where all warnings are?

Comment (fpetrogalli) [reply CMakeLists.txt:1494]: I'll mark this as done, the warnings are in the same place where the status messages were. Let me know if you want me to move them somewhere else.

File: CMakeLists.txt (hunk 8 above)
Comment (fpetrogalli) [reply CMakeLists.txt:548]: Hum - `OPENCV_DISABLE_THREAD_SUPPORT` does not seem to be visible in the openjpeg folder.

I get the following message when trying to build for baremetal after moving ` set(OPJ_USE_THREAD OFF)` in openjpeg:

```
/home/frapet01/projects/opencv/opencv/3rdparty/openjpeg/openjp2/thread.c: In function 'opj_thread_join':
/home/frapet01/projects/opencv/opencv/3rdparty/openjpeg/openjp2/thread.c:423:5: warning: implicit declaration of function 'pthread_join'; did you mean 'opj_thread_join'? [-Wimplicit-function-declaration]
  423 |     pthread_join(thread->thread, &status);
      |     ^~~~~~~~~~~~
      |     opj_thread_join
```

Would setting `set(OPJ_USE_THREAD OFF)` in the toolchain file of the baremetal platform OK? Or is there a more canonical way to do this?

File: modules/core/src/async.cpp (hunk 10 above)
Comment (fpetrogalli) [reply modules/core/src/async.cpp:208]: I started providing stubs for the API of `cv::AsyncArray`, and it turned out I also need to provide stubs for `cv::AsyncPromise`. The approach I followed was to remove the `AsyncArray::Impl` definition and uses. This however caused writing many `#ifundef / #else / #endif` cases, which doesn't seem to fit very well with the goal of miniming such cases.

Would it be OK if instead of providing stubs for `cv::AsyncArray` and `cv::AsyncPromise`, I would be providing  a stub class `AsyncArray::Impl` that would expose stubs only for the methods used by `cv::AsyncArray` and `cv::AsyncPromise`? This solution would still provide useful exception message, and would require just a couple of `#ifundef / #else / #endif` cases.

File: modules/core/src/system.cpp (hunk 11 above)
Comment (fpetrogalli) [reply modules/core/src/system.cpp:1518]: Done.

File: CMakeLists.txt (hunk 7 above)
Comment (fpetrogalli) [reply CMakeLists.txt:519]: Done. I'll move the FILESYSTEM one separately, though.

Comment (fpetrogalli) [reply CMakeLists.txt]: With the addition of #20137 , I consider this done.

**Comment by fpetrogalli** [comment]
@alalek  - thank you for the review. The baremetal builds are still running correctly even after the changes you requested, this is very nice. 

As a matter of fact, I need more feedback from you about https://github.com/opencv/opencv/pull/19985#discussion_r636818788, as I am not 100% sure what I have done is the right thing.

Thank you again,

Francesco

**Comment by fpetrogalli** [comment]
@alalek  - please note that, once you accept this patch, I think that the commits in this patch should be squashed before merging it into master.

File: CMakeLists.txt [hunk 12]
```diff
@@ -1458,6 +1491,15 @@ ocv_build_features_string(parallel_status EXCLUSIVE
   ELSE "none")
 status("")
 status("  Parallel framework:" "${parallel_status}")
+if (OPENCV_DISABLE_THREAD_SUPPORT)
+  status("" "Multi thread code explicitly disabled with OPENCV_DISABLE_THREAD_SUPPORT.")
+  if(HAVE_PTHREADS_PF OR HAVE_HPX OR HAVE_OPENMP OR HAVE_GCD OR HAVE_CONCURRENCY)
+    message(WARNING "Not all parallel frameworks have been disabled (using ${parallel_status}).")
```
Comment (asmorkalov) [inline CMakeLists.txt:1497]: I propose use FATAL message for the case, because `OPENCV_DISABLE_THREAD_SUPPORT` disables Mutex and other sync primitives and makes parallel implementations unsafe.

File: cmake/OpenCVFindLibsGUI.cmake [hunk 13]
```diff
@@ -34,6 +34,41 @@ if(WITH_QT)
...
+    ocv_append_build_options(HIGHGUI GTHREAD)
+  endif()
+  if(WITH_OPENGL AND NOT HAVE_GTK3)
+    ocv_check_modules(GTKGLEXT gtkglext-1.0)
+    if(HAVE_GTKGLEXT)
+      ocv_append_build_options(HIGHGUI GTKGLEXT)
+    endif()
+  endif()
+endif()
+
```
Comment (asmorkalov) [inline cmake/OpenCVFindLibsGUI.cmake:71]: It's not relevant to threading.

File: CMakeLists.txt [hunk 14]
```diff
@@ -523,6 +524,33 @@ if(OPENCV_DISABLE_FILESYSTEM_SUPPORT)
   add_definitions(-DOPENCV_HAVE_FILESYSTEM_SUPPORT=0)
 endif()
 
+if(OPENCV_DISABLE_THREAD_SUPPORT)
+  add_definitions(-DOPENCV_DISABLE_THREAD_SUPPORT)
```
Comment (asmorkalov) [inline CMakeLists.txt:528]: The option is not exported to `OpenCV-config.cmake` and other build configs. All methods in header files will use another pre-processor branch and break user code.

**Review by asmorkalov — CHANGES_REQUESTED** [review]

File: modules/ts/src/ts_gtest.cpp [hunk 15]
```diff
@@ -1067,6 +1067,7 @@ class GTEST_API_ UnitTestImpl {
   void AddTestInfo(Test::SetUpTestCaseFunc set_up_tc,
                    Test::TearDownTestCaseFunc tear_down_tc,
                    TestInfo* test_info) {
+#ifndef OPENCV_DISABLE_THREAD_SUPPORT
     // In order to support thread-safe death tests, we need to
     // remember the original working directory when the test program
```
Comment (alalek) [inline modules/ts/src/ts_gtest.cpp:1072]: How is directory related to disabled threads?
Suggesting to revert this change.

File: modules/ts/src/ts_perf.cpp [hunk 16]
```diff
@@ -1359,8 +1359,10 @@ bool TestBase::next()
         lastActivityPrintTime = 0;
         metrics.clear();
     }
-
+    
+#ifdef OPENCV_HAVE_THREAD_SUPPORT
     cv::theRNG().state = param_seed; //this rng should generate same numbers for each run
+#endif
```
Comment (alalek) [inline modules/ts/src/ts_perf.cpp:1365]: We still need to reset RNG state to initial value.
There is nothing about the threading.

All TLS containers must work, with behavior of regular variables.

File: CMakeLists.txt (hunk 14 above)
Comment (fpetrogalli) [reply CMakeLists.txt:528]: I am happy to do the export of the variable but I am not sure I understand the issue.

Also, how can I export `OPENCV_DISABLE_THREAD_SUPPORT` in `OpenCV-config.cmake`? I have looked a bit into `cmake/OpenCVGenConfig.cmake` but I am missing the mechanism that does this export.

File: modules/ts/src/ts_gtest.cpp (hunk 15 above)
Comment (fpetrogalli) [reply modules/ts/src/ts_gtest.cpp:1072]: Right - this is something that should probably be guarded by the `OPENCV_HAVE_FILESYSTEM_SUPPORT`. I have removed it from this PR, it might show up with a new guard in #19921

**Comment by fpetrogalli** [comment]
@alalek @asmorkalov  - Thank you for the review. Apologies for the delay in getting back to you but I was OoO.

Francesco

**Comment by fpetrogalli** [comment]
Hi @alalek  - thank you for the patch that you have added on top of my changes. Extracting the cmake commands in a separate file and reducing the number of `#ifdef` conditions make totally sense. The code looks much better now!

Unfortunately, the changes are breaking the build for the `aarch64-baremetal` branch at #19921 . I'll add my own commits on top of your to get #19921 to build with the changes on top of #19985 .

While I am at that, I will also merge all the commits of #19985 marked with "to squash" into a single commit (I won't include your commit in the squash).

Francesco

**Comment by fpetrogalli** [comment]
@alalek @asmorkalov  - gentle ping!

Francesco

**Comment by alalek** [comment]
Patch is rebased to resolve conflict in `cmake/OpenCVCompilerOptions.cmake` (near `add_extra_compiler_option(-pthread)` condition)

**Comment by fpetrogalli** [comment]
@alalek , thank you for looking into this code again.

> I'm unable to check #19921 (proposed toolchain requires AARCH64 host - I don't have such configured setup for development), please take a look on your side.

#19921 is still to considered a WIP. There is no need to review it for now. I'd rather have this PR (#19985 ) done and dusted before asking you to look in the baremetal one. We have CI that depends on #19921 , for now I just keep updating it to make our CI happy again when things start failing because of updates or merges.

File: CMakeLists.txt [hunk 17]
```diff
@@ -520,6 +521,10 @@ if(ENABLE_IMPL_COLLECTION)
   add_definitions(-DCV_COLLECT_IMPL_DATA)
 endif()
 
+if (OPENCV_DISABLE_THREAD_SUPPORT)
+  set(WITH_GSTREAMER OFF)
+endif()
```
Comment (alalek) [inline CMakeLists.txt:526]: Why do we need this?

It should be handled in `cmake/vars/OPENCV_DISABLE_THREAD_SUPPORT.cmake`: https://github.com/opencv/opencv/pull/19985/files#diff-87b6de92e07d14f041e4b331f8936a5d5a4f2f8e499aa71bc3967de40cff9960

Please avoid any specific changes in root CMakeLists.txt - it should not manage details of 3rdparty components (otherwise we would get unsupportable huge mess)

Comment (fpetrogalli) [reply CMakeLists.txt:526]: > It should be handled in cmake/vars/OPENCV_DISABLE_THREAD_SUPPORT.cmake
I agree, sorry, I keep forgetting about this file and I end up running in the same issue everytime I was recompiling after changing branch because...

> Ensure that you have clean CMake cache (after code update)
... I was not aware of this trick! Thank you. I am going to remove the last commit  with this change.

**Review by alalek — APPROVED** [review]
Thank you for contribution :+1:

**Comment by fpetrogalli** [comment]
Thank you @alalek ! Now I'll clean up #19921 and get it ready for review!


Francesco
